src/server/services/records.js

```diff
@@ -102,7 +103,19 @@ async function recordsForUpload (upload) {
 
     // each row in the input sheet becomes a unique record
     for (const row of rows) {
-      records.push({ type, subcategory, upload, content: row })
+      const formattedRow = {}
+      Object.keys(row).forEach(fieldId => {
+        let value = row[fieldId]
+        for (const formatter of typeRules[fieldId].persistentFormatters) {
+          try {
+            value = formatter(value)
+          } catch (e) {
+            console.log(`Persistent formatter failed to format value ${value} with error:`, e)
+          }
+        }
+        formattedRow[fieldId] = value
```

Comment:
Totally optional, but since we have lodash in the project, you could try leveraging the `_.flow` pattern here!

```suggestion
try {
  const value = _.flow(typeRules[fieldId].persistentFormatters)(row[fieldId])
  formattedRow[fieldId] = value
} catch (e) {
  console.log(`Persistent formatter failed to format value ${value} with error:`, e)
}
```

I guess one big difference is that your approach supports skipping an individual broken formatter.  I'm not sure if that's likely to be more or less confusing in practice.

Reply:
I do like how the `flow` code reads, but I think I prefer the functionality of not short-circuiting if one is broken. Mostly because this way we know that any broken formatter will write an error message out (which will be even more useful once we get sentry hooked up)